### ISSUE TITLE
Vulkan backend does not reload model into vram if kicked out by other uses. Massive slowdown.

### ISSUE THREAD

**Issue by superkuh** [ggerganov/llama.cpp#5380]
I recently switched to the vulkan backend. It's great and faster than clBLAST. Unfortunately it has some quirks. With clBLAST and older llama.cpp if some other program used all the GPU vram and kicked the model out llama.cpp would automatically reload the model and continue at full speed. The reloading of the model is easily visible in radeontop.

But with the new vulkan llama.cpp builds if another program uses all the vram, kicks out most of the model weights, then llama.cpp never reloads them and the speed drops from 10 tokens/s to 1 token/s. Looking at my GPU ram usage with radeontop I can see vram usage never goes above ~1GB usage and never reloads the model. This remains the broken state till ./server is killed and restarted. At which point radeontop shows it loaded the full model into vram again.

The way I trigger this bug is to run the ./llava binary and load a multimodal model into vram kicking the ./server data out of GPU VRAM. But any secondary process using most vram should trigger it.

**Comment by 0cc4m** [comment]
Interesting. I'm not sure why this would happen or what it's doing exactly. But this isn't something I have control over, in the Vulkan backend I just allocate device memory. If that gets moved to RAM, that's a driver operation and the driver should move it back once VRAM is available again. I don't think I can solve this from within the program.

**Comment by superkuh** [comment]
Okay. Dang.

I've talked to some smart OFTC #radeon people and they think it might be fixable through compiling the amdgpu driver myself with these 2 patches and replacing the Debian 11 version, 

https://patchwork.freedesktop.org/series/129196/ "Try to fill up VRAM as well by setting the busy flag on GTT allocations. This fixes the issue that when VRAM was evacuated for suspend it's never filled up again unless the application is restarted."

That's a bit beyond my capabilities but I'm noting it here for anyone else that has the problem.

**Comment by gjuoun** [comment]
Same here (rx570 8gb), please keep it open

**Comment by 0cc4m** [comment]
> Same here (rx570 8gb), please keep it open

Unless you figure out how to influence this behaviour from the client side, this is a driver issue and nothing we can do anything about in this library.

**Comment by 0cc4m** [comment]
@gjuoun Yeah, look at my earlier comment: https://github.com/ggerganov/llama.cpp/issues/5380#issuecomment-1932667035

**Comment by netrunnereve** [comment]
I'm actually seeing something similar with Vulkan, two GPUs, and the amdgpu driver. In my case the first GPU works fine with everything in VRAM but the second GPU sends everything to GTT and inference becomes super slow. What's different from your case is that this happens immediately when the model is loaded with lots of free VRAM available.

I managed to solve this by turning off amdgpu runtime power management (kernel flag amdgpu.runpm=0). The first GPU is always on since it's plugged into my monitors and shows my desktop, but the second one ends up being suspended by Linux as it doesn't do anything most of the time. Apparently amdgpu will send everything to GTT until the GPU actually wakes up to run inference, and as we've seen here the driver isn't smart enough to copy everything back to VRAM again.

If I run something like `glxgears` on the second GPU before running llama.cpp everything gets sent to VRAM, so maybe our backend could be modified to give the GPU a bit of work before allocating memory. But I'm just rambling at this point and everything works fine with power management turned off.

**Comment by JeffKwasha** [comment]
I'm running Kubuntu 2 monitors. rocm 7.2.3 on 30.20.1

I've seen this with llama.cpp
r9700 32GB VRAM (~31 in use), 48GB system ram (~25 in use) - submit a couple sentences in opencode and suddenly the whole system hangs.

I ssh in 'server systemctl stop --user llama-server'

A minute later it unfreezes and I get 3 popups that OOM has killed brave etc. and my swap (5G) is full.

The only thing big enough to swamp RAM was the 22GB Qwen (and probably some kv cache)

**Comment by nzbrian** [comment]
I'm seeing this in Windows 11, Vulkan, using a headless RX 7900 XTX. Watching the VRAM slowly unload to RAM and then back again after any pause. This was hard to work out, and plugging in a fake monitor fixed it instantly so thank you.

**Comment by CircuitGuy** [comment]
> amdgpu.runpm=0

Almost 2 years later @netrunnereve saved me with this one. Sure enough, if the GPU sleeps due to no monitor, it dumps all of VRAM to SysRAM, potentially crashing the system.
